File: benchmark2.py
SIZE = 50000
EXPECTED_CHECKSUM = 41667916675000

def main():
    # Initialize arrays
    A = [i for i in range(SIZE)]
    B = [SIZE - i for i in range(SIZE)]
    C = [0 for _ in range(SIZE)]

    # Precompute prefix sums of B
    prefix_sum = [0] * SIZE
    prefix_sum[0] = B[0]
    for i in range(1, SIZE):
        prefix_sum[i] = prefix_sum[i - 1] + B[i]

    # Compute C[i] using prefix sums
    for i in range(SIZE):
        C[i] = prefix_sum[A[i]]

    # Compute checksum
    checksum = sum(C)

    # Verify checksum
    if checksum == EXPECTED_CHECKSUM:
        print("Checksum =", checksum)
        return 0
    else:
        print("Error: checksum mismatch! Got", checksum, "expected", EXPECTED_CHECKSUM)
        return 1

if __name__ == "__main__":
    main()

File: benchmark2.c
#include <stdio.h>
#include <stdlib.h>

#define SIZE 50000
#define EXPECTED_CHECKSUM 41667916675000LL

int main() {
    static long long A[SIZE];
    static long long B[SIZE];
    static long long C[SIZE];
    long long checksum = 0;

    // Initialize arrays
    for (int i = 0; i < SIZE; i++) {
        A[i] = i;
        B[i] = SIZE - i;
        C[i] = 0;
    }

    // Perform computation
    for (int i = 0; i < SIZE; i++) {
        long long sum = 0;
        for (int j = 0; j <= A[i]; j++) {
            sum += B[j];
        }
        C[i] = sum;
    }

    // Compute checksum
    for (int i = 0; i < SIZE; i++) {
        checksum += C[i];
    }

    // Verify checksum
    if (checksum == EXPECTED_CHECKSUM) {
        printf("Checksum = %lld\n", checksum);
        return 0;
    } else {
        fprintf(stderr, "Error: checksum mismatch! Got %lld, expected %lld\n", checksum, EXPECTED_CHECKSUM);
        return 1;
    }
}